### PR TITLE
#16 render battles

### LINKED ISSUE #16: Render battles
AC

- [x] render a complete battle

### PR DESCRIPTION
**Pull request by proSingularity**
FIxes #16

@knoffi Please review. If you have any questions always feel free to ask via comment in the PR. Also: I recommend commit-wise review. 
Hint: Under the Github tab `Files changed` in this PR, you can see each file. Atop the file there's a checkboxbutton `Viewed` that's super usefl. 

Changes:

- Complete battle is rendered with some animations (Fixes #16)
- Migrate to [Phaser3](https://phaser.io/phaser3)
  - Reason: Making animations with Vanilla JS is not feasable for me.
  - includes lots of boilerplate, e.g. testing (mocha, instanbul/nyc), listing (eslint, editor config, prettier)
- Page redesign with Gameboy Advance and battle rendered on its screen
  - Reason: Because it's cool =)
- added `db.json` under `/web/db.json`
  - Reason: To simpliify frontend development, I often make changes to my `db.json`. But the webserver needs the file somewhere under `/web/`. Otherwise it cannot serve it. The webserver cannot do `../` to mitigate unwanted file access attacks.
  - Added a copy for now to avoid merge conflicts. Will backmerge the changes from `master` on next occassion.

Known Issues: 
- cannot restart a battle / start a 2nd battle
  - TODO @proSingularity: create issue
- crashes when the `Start Pokemon Battle` gets clicked twice
  - should instead start 2nd battle
  - TODO @proSingularity: create issue to mitigate click-spamming
- ~~some pokemon are slightly too large, E,g, venomoth and ninetails~~
  - mitigated by making team at the top smaller (also indicates distance from camera). As a Pokemon attacks and moves closer to the camera it scales simultaneously.
  - @knoffi If you think this obfuscates the PR, I can just pull it out of the PR and create a 2nd PR. i'm much in favor of smaller PRs. Sorry for this infrastructure-heavy, large PR. 
- due to the gameboy advance design, **mobile compatibility is lost**
  - possibly mitigated by forcing landscape mode, or using a `@media` query to completely hide the gameboy and only display the Phaser canvas (the canvas should be small enough). Personally, I currently favor hiding the gameboy on mobile for it's simplicity. Further investigation needed.
  - TODO @proSingularity: open discussion with @knoffi 
- Building the webapp for deployment untested and probably broken
  - TODO @proSingularity: open issue

### CODE REVIEW

File: web/src/scenes/StandbyScene.ts [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+        // eslint-disable-next-line @typescript-eslint/no-misused-promises
+        startBattle.addEventListener("click", async () => {
+            this.add.image(0, 0, "gbastartscreen").setOrigin(0).setScale(1.01); // remove 1-3px thin grey line on the right
+            const beforeFetch = this.time.now;
+            const api = isProd ? url : devUrl;
+            const data = await fetch(api).then(
+                (x) => x.json() as Partial<ApiRes>
+            );
+            assertApiRes(data);
+            // NOTE: don't use scene.add without true + scene.start here, otherwise phaser will just ignore the start call for some reason. Bug?
```
Comment (knoffi) [inline web/src/scenes/StandbyScene.ts:89]: I think, Phaser is doing this on purpose. https://rexrainbow.github.io/phaser3-rex-notes/docs/site/scenemanager/ 

But I did not know about that until 5 minutes ago. : )

File: web/src/scenes/StandbyScene.ts [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+        typeof x.backSprite !== "string" ||
+        typeof x.frontSprite !== "string" ||
+        typeof x.name !== "string"
+    ) {
+        return false;
+    }
+    return true;
+};
+
+function assertApiRes(x: undefined | Partial<ApiRes>): asserts x is ApiRes {
```
Comment (knoffi) [inline web/src/scenes/StandbyScene.ts:30]: Very nice

File: web/src/scenes/StandbyScene.ts (hunk 1 above)
Comment (proSingularity) [reply web/src/scenes/StandbyScene.ts:89]: Thanks for the link, @knoffi. Was able to figure out another way to make it work that is very close to what I originally had. I must have made a mistake last time.
```ts
this.scene.add(Scenes.Main, MainScene);
this.scene.start(Scenes.Main, data);
```
Will leave the current code though as it does the same. Just removed the comment.
(PS: You can close the discussion if you think this is fine now. This is helpful to understand the state of the PR/review.)

**Comment by proSingularity** [comment]
@knoffi solved merge conflicts. Note the two manually written `FIX_MERGE` commits. 
Why did this occur?
Due to moving `/web/index.html` and all other contents of `/web` into `/src/main/resources/static` git thought that it should also move all my newly created files to the `static` directory. From a build perspective however, this is not what we want, because a separate `build` step is now necessary for the frontend (`build` step is to be implemented as a follow-up).

I definitely didn't expect git to do this. There might have been some way to avoid such git behavior but i'm not aware of how to do that.

Anyway, what I did was move all files back to `/web` manually.
An attempt to avoid git moving all files automatically failed, so this is the lesser of the two evils.

File: web/src/scenes/StandbyScene.ts [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+            "Invalid response from backend. Team contains non-IPokemon"
+        );
+    }
+}
+
+const devUrl = "./db.json";
+const url =
+    "https://raw.githubusercontent.com/knoffi/pokeBattleSim/main/db.json";
+
+export class StandyScene extends Scene {
```
[the remark this answers is not in the document]
Comment (proSingularity) [reply web/src/scenes/StandbyScene.ts:53]: fixed

**Review by knoffi — APPROVED** [review]
Wonderful work, thank you very much. I can not wait to test it.

**Pull request merged**